src/backend/distributed/replication/multi_logical_replication.c

```diff
@@ -193,30 +197,47 @@ LogicallyReplicateShards(List *shardList, char *sourceNodeName, int sourceNodePo
 
 	PG_TRY();
 	{
-		/*
-		 * We have to create the primary key (or any other replica identity)
-		 * before the initial COPY is done. This is necessary because as soon
-		 * as the COPY command finishes, the update/delete operations that
-		 * are queued will be replicated. And, if the replica identity does not
-		 * exist on the target, the replication would fail.
-		 */
-		CreateReplicaIdentity(shardList, targetNodeName, targetNodePort);
-
 		/* set up the publication on the source and subscription on the target */
 		CreateShardMovePublications(sourceConnection, replicationSubscriptionList,
 									tableOwnerIds);
-		CreateShardMoveSubscriptions(targetConnection, sourceNodeName, sourceNodePort,
-									 superUser, databaseName, tableOwnerIds);
+		char *snapshot = CreateShardMoveSubscriptions(
+			sourceConnection,
+			targetConnection,
+			sourceReplicationConnection,
+			databaseName,
+			tableOwnerIds);
 
 		/* only useful for isolation testing, see the function comment for the details */
 		ConflictOnlyWithIsolationTesting();
 
+		WorkerNode *sourceNode = FindWorkerNode(sourceNodeName, sourceNodePort);
+		WorkerNode *targetNode = FindWorkerNode(targetNodeName, targetNodePort);
+		CopyShardsToNode(sourceNode, targetNode, shardList, snapshot);
+
 		/*
-		 * Logical replication starts with copying the existing data for each table in
-		 * the publication. During the copy operation the state of the associated relation
-		 * subscription is not ready. There is no point of locking the shards before the
-		 * subscriptions for each relation becomes ready, so wait for it.
+		 * We can close this connection now, because we're done copying the
+		 * data and thus don't need access to the snapshot anymore. The
+		 * replication slot will still be at the same LSN, because the
+		 * subscriptions have not been enabled yet.
 		 */
+		CloseConnection(sourceReplicationConnection);
+
+		/*
+		 * We have to create the primary key (or any other replica identity)
+		 * before the update/delete operations that are queued will be
+		 * replicated. Because if the replica identity does not exist on the
+		 * target, the replication would fail.
+		 *
+		 * So we it right after the initial data COPY, but before enabling the
+		 * susbcriptions. We do it at this latest possible moment, because its
+		 * much cheaper to build an index at once than to create it
+		 * incrementally. So this way we create the primary key index in one go
+		 * for all data from the initial COPY.
+		 */
+		CreateReplicaIdentity(shardList, targetNodeName, targetNodePort);
+
+		EnableShardMoveSubscriptions(targetConnection, tableOwnerIds);
+
```

Comment:
While going over this code again, I realized an issue with foreign keys between distributed tables (unrelated to this PR):

We currently create foreign keys between reference tables in CreatePostLogicalReplicationDataLoadObjects. However, since we could have multiple subscriptions, the changes might not be synchronized and foreign keys could be violated on the target.

That might also cause issues with truncate:
https://www.postgresql.org/docs/current/logical-replication-restrictions.html

In addition, we currently check the whole foreign key duration creation even though we know it to be valid on the source.

We should probably handle foreign keys between distributed tables in the same way as we do foreign keys to reference tables: Create them all the way at the end after the last catchup and and skip validation.

Reply:
Created a separate issue for this: https://github.com/citusdata/citus/issues/6141